kythe/rust/indexer/src/providers.rs

```diff
@@ -0,0 +1,152 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::KytheError;
+use analysis_rust_proto::*;
+use positioned_io;
+use rc_zip::{prelude::*, Archive, EntryContents};
+use std::fs::File;
+use std::io::Read;
+
+/// A trait for retrieving files during indexing.
+pub trait FileProvider {
+  /// Checks whether a file exists and that the contents is available.
+  ///
+  /// This function takes a file hash and returns whether
+  /// the file exists and that the contents can be requested.
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a [KytheError::KzipFileError] will
+  /// be returned.
+  fn exists(&mut self, file_hash: &str) -> bool;
+
+  /// Retrieves the string contents of a file.
+  ///
+  /// This function takes a file hash and returns the string contents of the
+  /// file.
+  ///
+  /// # Errors
+  ///
+  /// If the file does not exist, a
+  /// [FileNotFoundError][KytheError::FileNotFoundError] will be returned.
+  /// If the file can't be read into a String, a
+  /// [FileReadError][KytheError::FileReadError] will be returned.
+  fn contents(&mut self, file_hash: &str) -> Result<String, KytheError>;
+}
+
+/// A [FileProvider] that backed by a .kzip file.
+pub struct KzipFileProvider {
+  file: File,
+  zip_archive: Archive,
+}
+
+impl KzipFileProvider {
+  /// Create a new instance using a valid [File] object of a kzip
+  /// file.KytheError
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a
+  /// [KzipFileError][KytheError::KzipFileError] will be returned.
+  pub fn new(file: File) -> Result<Self, KytheError> {
+    let zip_archive = match file.read_zip() {
+      Ok(zip_archive) => zip_archive,
+      Err(_e) => return Err(KytheError::KzipFileError),
+    };
+    Ok(Self { file, zip_archive })
+  }
+
+  /// Retrieve the Compilation Units from the kzip.
+  ///
+  /// This function will attempt to parse the proto files in the `pbunits`
+  /// folder into Compilation Units, returning a Vector of them.
+  ///
+  /// # Errors
+  ///
+  /// If a file cannot be parsed, a
+  /// [ProtobufParseError][KytheError::ProtobufParseError] will be returned.
+  pub fn get_compilation_units(&self) -> Result<Vec<CompilationUnit>, KytheError> {
+    let mut compilation_units = Vec::new();
+    for entry in self.zip_archive.entries() {
+      // Protobuf files are in the pbunits folder
+      if entry.name().contains("/pbunits/") {
+        // Read the file into a compilation unit protobuf
+        match entry.contents() {
+          EntryContents::File(f) => {
+            let mut entry_reader = f
+              .entry
+              .reader(|offset| positioned_io::Cursor::new_pos(&self.file, offset));
+            let bundle = match ::protobuf::parse_from_reader::<CompilationBundle>(&mut entry_reader)
```

Comment:
Does just `protobuf::parse_from_reader` (without the leading colons) work? I don't often see that `::foo` syntax outside of macros.

Reply:
I think I copied this from one of the examples. I'll try removing it and see if it works.